Move IFrame-host container creation into IFrame from outside

heliocliu opened the pull request:
Fixes #4099 (and maybe #1333?)

Move container creation into the IFrame from outside.  Main changes include:
-Decouple `resolvedUrl` from the document service/document service factory instances and instead provide it during creation like the other non-iframe implementations.  This involves some API changes as well to work with `postMessage`, and also means some wrapping layers like `IFrameDocumentServiceProxyFactory ` can be removed.
-Add a `UrlResolver` proxy that works in the same way as the existing document service proxy.  This also involves changing how we expose/wrap objects on the postMessage boundary using comlink, because only a single object can be exposed this way
-Add a `container` proxy for the outer to use to communicate with the container created inside the iframe
-Use thin getter lambdas wrapped with `Comlink.proxy` when passing non-primitives across the `postMessage` boundary instead of creating real proxies.  This works for the case here (aside from the logger), but longer term real inner/outer proxy pairs will likely be necessary because `postMessage` permeability is affected by internal implementation.
-Create the container within the IFrame using the same `getContainer` API from within `getTinyliciousContainer` that was used outside the IFrame, but leave it to the outer host to call `attach` on the created container.  This also involves exporting `InsecureTinyliciousUrlResolver` because we need to use `InnerDocumentService` instead of the `RouterliciousDocumentService` that `getTinyliciousContainer` uses internally.
-Use existing comlink for outer to send messages to inner, but use MessageChannels for inner to communicate with outer (because comlink doesn't support two way).  This needs to be unified all into MessageChannel or something else moving forward.

heliocliu committed:
initial driver changes

-copy implementations for createContainer
-update comlink dependencies

heliocliu committed:
remove persisted url from driver

-force the url to be provided when creating the driver to match what other drivers currently do (and allow the creation of multiple drivers)

heliocliu committed:
remove iframedocumentserviceproxyfactory

Remove this class now that the resolved url has been decoupled from the iframe driver (and there's no reason to have a factory that creates the other factory)

heliocliu committed:
move creation into iframe

Move data store creation into the iframe, triggered from outside the iframe
-Use a div in the html and create the iframe in code
-Expose a loadFluidObject method directly on the iframe window from Inner
-Export InsecureTinyliciousUrlResolver for the purposes here (temporary), needed to continue using the getContainer method inside the iframe because we also need to use InnerDocumentServiceFactory there
-Change Outer to call the exposed loadFluidObject method, and load the log output/data object afterward (as opposed to setting up the log div and outer div then waiting to populate it later)

heliocliu committed:
Merge branch 'main' into MoveContainerCreation

heliocliu commented on packages/drivers/iframe-driver/src/proxyUtils.ts:
```diff
@@ -0,0 +1,10 @@
+
+import * as Comlink from "comlink";
+
+export const MakeThinProxy = <T>(obj: T): () => Promise<T> => {
```
I doubt we'll want to keep this around longer term but it's here mainly to make things a bit easier to read

heliocliu committed:
policy check

heliocliu committed:
remove unused proxyloader

anthony-murphy commented on examples/hosts/iframe-host/src/inner.ts:
```diff
@@ -24,26 +28,36 @@ async function getFluidObjectAndRender(container: Container, div: HTMLDivElement
-        documentServiceFactory,
-        codeLoader,
-    });
+    const urlResolver = new InsecureTinyliciousUrlResolver();
```
>InsecureTinyliciousUrlResolver [](start = 28, length = 30)

the inner should never use real drivers. it should proxy everything out to the outer driver

anthony-murphy commented on examples/hosts/iframe-host/src/inner.ts:
```diff
@@ -24,26 +28,36 @@ async function getFluidObjectAndRender(container: Container, div: HTMLDivElement
-    const url = documentServiceFactory.resolvedUrl.url;
-    const documentId = decodeURIComponent(url.replace("fluid://localhost:3000/tinylicious/", ""));
-    const container = await loader.resolve({ url: documentId });
+    const container = await getContainer(
```
>getContainer [](start = 28, length = 12)

i don't think this is the right thing to use either, as we should support using an actual code loader here. i'd prefer this use the raw pieces, like loader

anthony-murphy commented on examples/hosts/iframe-host/src/outer.ts:
```diff
@@ -31,40 +30,37 @@ const getTinyliciousUrlResolver =
     const documentServiceFactory = new RouterliciousDocumentServiceFactory(tokenProvider);
 
+    const module = { fluidExport: TodoContainer };
+    const codeLoader = { load: async () => module };
```
the outer should never need the real code, the whole point is to isolate the real code in the iframe

anthony-murphy commented on examples/hosts/iframe-host/src/outer.ts:
```diff
@@ -83,25 +79,40 @@ async function getFluidObjectAndRender(container: Container, div: HTMLDivElement
+): Promise<void> {
+    const logDiv = document.getElementById(logDivId) as HTMLDivElement;
 
-    const container = await getTinyliciousContainer(
```
>container [](start = 10, length = 9)

this was fine, as this function is really just for testing to ensure we can had a div version of the same conntainer, and gives us another client to connect to the container in the iframe. it shouldn't reuse anything from the iframe case.

anthony-murphy commented on examples/hosts/iframe-host/src/inframehost.ts:
```diff
@@ -90,29 +38,19 @@ export class IFrameOuterHost {
+    }
 
-        return container;
+    public async getContainerForRequest(request: IRequest): Promise<Container> {
```
this shouldn't exist here, as the iframehost should never load or be exposed to the inner code. i would have expected a createDetached here that returned some kind of proxy to a detached container inside the iframe on which the outer could call attach

anthony-murphy replied:
to start with, it's probably easier to just have the inner just create detached, and call attach, which get's proxied to the outer. what i call out above could come i follow up

---
In reply to: [527049230](https://github.com/microsoft/FluidFramework/pull/4404#discussion_r527049230) [](ancestors = 527049230)

heliocliu replied:
I created a container proxy layer and moved the attach call to the outer host.  Wasn't too clear on what you meant in the second suggestion though, because calling an `attach` that's been proxied to the outer host would mean the real container exists in outer, which didn't seem like what we're going for

anthony-murphy commented on packages/drivers/iframe-driver/src/outerDocumentServiceFactory.ts:
```diff
@@ -102,51 +119,72 @@ export class DocumentServiceFactoryProxy implements IDocumentServiceFactoryProxy
 
     // TODO: Issue-2109 Implement detach container api or put appropriate comment.
+    // Currently this is mostly copied from RouterliciousDocumentServiceFactory
     public async createContainer(
```
i don't understand why anything is copied, shouldn't this just be using the real documentServiceFactory that is passed in?

anthony-murphy replied:
this shouldn't be assuming anything about the server. it should used the public interfaces on our existing driver apis. if that doesn't work, we should add new apis or modify existing apis

---
In reply to: [527056264](https://github.com/microsoft/FluidFramework/pull/4404#discussion_r527056264) [](ancestors = 527056264)

anthony-murphy commented on packages/drivers/iframe-driver/src/outerDocumentServiceFactory.ts:
```diff
@@ -3,26 +3,29 @@
  */
 
+import { parse } from "url";
+import Axios from "axios";
```
>Axios [](start = 7, length = 5)

this shouldn't be making requests on it's own

anthony-murphy commented on examples/hosts/iframe-host/src/inner.ts:
```diff
@@ -3,8 +3,12 @@
+import {
+    getContainer,
+    InsecureTinyliciousUrlResolver,
+ } from "@fluidframework/get-tinylicious-container";
```
the inner shouldn't have any real drivers, as we want to keep the iframe code isolated, and not have to give it tokens or anything

heliocliu committed:
pr feedback pt1

-OuterDocumentServiceFactory refactoring to correctly call real factory instead of copying code lol
-Extract UrlResolver out of the iframe and create proxy objects etc.
-Restructure exposed proxy on the window to be a combined object of all proxes and rely on the host to compose the combined proxy object (instead of handling it in the "outer" objects).  Necessary for comlink because it doesn't handle exposing multiple different objects on the same window and only fails with cryptic runtime errors
-Unfactor getContainer in inner

heliocliu committed:
pr feedback2

-Proxy out container attach in the createNew case to the outer host.  Create a ContainerProxy object to do so.  Gvien the outer host then has access to the container, it doesn't need to do anything with a loader to get the fluid object anymore.

heliocliu committed:
Merge branch 'main' into MoveContainerCreation

anthony-murphy commented on examples/hosts/iframe-host/src/inner.ts:
```diff
@@ -24,26 +28,62 @@ async function getFluidObjectAndRender(container: Container, div: HTMLDivElement
+        // We're not actually using the code proposal (our code loader always loads the same module regardless of the
+        // proposal), but the Container will only give us a NullRuntime if there's no proposal.  So we'll use a fake
+        // proposal.
+        container = await loader.createDetachedContainer({ package: "no-dynamic-package", config: {} });
```
>container = await loader.createDetachedContainer({ package: "no-dynamic-package", config: {} }); [](start = 8, length = 96)

not necessary in this PR, but we should create a follow up issue to support proxying this in from the loader

anthony-murphy commented on examples/hosts/iframe-host/src/inner.ts:
```diff
@@ -24,26 +28,62 @@ async function getFluidObjectAndRender(container: Container, div: HTMLDivElement
+
+export async function runInner(divId: string) {
+    // expose the entrypoints on the iframe window to load a Fluid object
+    (window as any).loadContainer = async (documentId, createNew) => {
```
>(window as any) [](start = 4, length = 15)

this shouldn't be necessary

heliocliu replied:
our compiler rules don't like us setting arbitrary properties like this :x it will be removed in the followup when it gets moved to use the postmessage api though

---
In reply to: [530567461](https://github.com/microsoft/FluidFramework/pull/4404#discussion_r530567461) [](ancestors = 530567461)

anthony-murphy commented on examples/hosts/iframe-host/src/inner.ts:
```diff
@@ -24,26 +28,62 @@ async function getFluidObjectAndRender(container: Container, div: HTMLDivElement
     });
 }
+
+export async function runInner(divId: string) {
```
not in this PR, but we should file an issue  to change this function to just setup a proxy, and let loading vs createNew be driven by the outer, similar to my comment above

anthony-murphy commented on examples/hosts/iframe-host/src/outer.ts:
```diff
@@ -45,29 +37,23 @@ export async function loadFrame(iframeId: string, logId: string) {
+        void (async () => {
+            // load the code inside the iframe but attach it here outside
+            const containerProxy: IContainerProxy =
+                await (iframe.contentWindow as any).loadContainer(documentId, createNew);
```
>await (iframe.contentWindow as any).loadContainer(documentId, createNew); [](start = 16, length = 73)

this shouldn't work, at least we shouldn't need it to work as we should support cross origin and sandbox iframes where we have no code access to the internal frame, only post message

heliocliu replied:
Trying to make the postmessage proxy API bidirectional with comlink doesn't "just work", so I think this and restricting the iframe permissions might be better suited in a followup

---
In reply to: [530570016](https://github.com/microsoft/FluidFramework/pull/4404#discussion_r530570016) [](ancestors = 530570016)

anthony-murphy replied:
this only works because you are not sandboxing the iframe's code e.g. the code is running on the outer page. this defeats the purpose of iframeing

---
In reply to: [530660933](https://github.com/microsoft/FluidFramework/pull/4404#discussion_r530660933) [](ancestors = 530660933,530570016)

anthony-murphy commented on examples/hosts/iframe-host/src/containerProxy.ts:
```diff
@@ -0,0 +1,35 @@
+
+export interface IContainerProxy {
+    attach(request: IRequest): Promise<void>;
+    request(request: IRequest): Promise<IResponse>;
```
we shouldn't proxy request, as request generally returns object which can't cross the iframe boundary. someday we may want a proxy for that stuff, but for now we should just punt on it

anthony-murphy commented on examples/hosts/iframe-host/src/containerProxy.ts:
```diff
@@ -0,0 +1,35 @@
+export interface IContainerProxy {
+    attach(request: IRequest): Promise<void>;
+    request(request: IRequest): Promise<IResponse>;
+    getQuorum(): Promise<IQuorum>;
```
>getQuorum(): Promise<IQuorum>; [](start = 4, length = 30)

originally we didn't need to proxy the quorum, as that part of the Container could live outside the iframe, as it didn't require any untrusted code, only the container code. i see why things like attach make that more complicated, as that will need to be proxied. we'll probably want a hybrid approach eventually. my intuition is that we'll need a full proxy for the detached case, but the load case won't need as much/any, as we can just use the Container

anthony-murphy replied:
we used the ProxyRuntime to get the full container on the outside. i'd like to keep support for that

---
In reply to: [530573659](https://github.com/microsoft/FluidFramework/pull/4404#discussion_r530573659) [](ancestors = 530573659)

anthony-murphy commented on examples/hosts/iframe-host/src/containerProxy.ts:
```diff
@@ -0,0 +1,35 @@
+    request(request: IRequest): Promise<IResponse>;
+    getQuorum(): Promise<IQuorum>;
+    on(event: string, listener: (...args: any[]) => void): Promise<void>;
+    once(event: string, listener: (...args: any[]) => void): Promise<void>;
```
i'm not sure we need to proxy these events. are they needed? we should be minimal here, and only handle what we need

anthony-murphy commented on examples/hosts/iframe-host/src/outer.ts:
```diff
@@ -83,25 +69,43 @@ async function getFluidObjectAndRender(container: Container, div: HTMLDivElement
 
-export async function loadDiv(divId: string) {
-    const div = document.getElementById(divId) as HTMLDivElement;
+async function loadOuterDataStoreAndLogDivs(
```
this is wrong. this shouldn't reuse anything from the iframe case. it should be a completely standalone, and just use the normal flow to load in the div. it is simply for testing to have a second client

heliocliu replied:
misunderstood the intention of the previous comment which is also what lead to the IContainerProxy bloat.  should be a lot more minimal now.

---
In reply to: [530576670](https://github.com/microsoft/FluidFramework/pull/4404#discussion_r530576670) [](ancestors = 530576670)

anthony-murphy commented on packages/drivers/iframe-driver/src/outerDocumentServiceFactory.ts:
```diff
@@ -43,6 +39,7 @@ const socketIOEvents = [
 
 export interface ICombinedDriver {
     clientId: string;
+    getResolvedUrl: () => Promise<IResolvedUrl>;
```
>IResolvedUrl [](start = 34, length = 12)

i'd rather not see this change combined. the resolved url can container credentials so we need to be really careful about proxying it

anthony-murphy commented:
It's really important to keep the principles of the change in mind. Specifically, we should never load, run, or directly access the loaded container runtime code outside of the iframe, and inside the iframe we should never have direct access to anything driver level, we should proxy everything that would need some level of credential to the outer keeping the inner fully isolated. it may be worthwhile to make the inner iframe fully sandboxed, and even disallow any connections from it. so, sandbox="allow-scripts" [https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe)

heliocliu committed:
remove access to outer resolvedUrl

heliocliu committed:
pr feedback

-Revert how the outer divs are loaded, back to loading with the proxyruntime for the log div, and using the regular non-iframe loading flow for the outer data store div
-Move loading the inner data store from the created container entirely back into the inner iframe, so that extra stuff doesn't need to be exposed on the container proxy (difference being that the fluid object now may be loaded into the inner div before it is attached), and clean up that now unused exposed API

heliocliu committed:
fix docs

heliocliu committed:
sandbox iframe

-Sandbox iframe with permission allow-scripts,allow-same-origin
-Add 2 way postMessage handling (sort of)

heliocliu commented on examples/hosts/iframe-host/src/outer.ts:
```diff
@@ -22,20 +27,21 @@ const getDocumentId = () => {
+    const iframeDiv = document.getElementById(iframeDivId) as HTMLIFrameElement;
+    const iframe = document.createElement("iframe");
+    iframe.src = "/inner.html";
+    // TODO: remove "allow-same-origin"
```
>// TODO: remove "allow-same-origin" [](start = 4, length = 35)

Not actually sure if it should be removed?

heliocliu committed:
bring back container proxy

bring back the container proxy to match how the other inner/outer proxies work

heliocliu committed:
Merge remote-tracking branch 'upstream/main' into MoveContainerCreation

anthony-murphy commented on examples/hosts/iframe-host/src/inner.ts:
```diff
@@ -24,26 +33,80 @@ async function getFluidObjectAndRender(container: Container, div: HTMLDivElement
+    }
+
+    await loadFluidObject(divId, container);
+    containers.set(container.id, container);
```
this is fine for now, but long term this should live in the iframe driver. 
Basically, the outer should call some create method, maybe passing an iframe, which returns a container proxy to the detached container in the iframe, which can then have attach called on it.
Ideally this is just a usage example, and doesn't really contain scenario implementation.

heliocliu replied:
As in the iframe-driver layer is responsible for keeping track like this of containers created inside the iframe?  `containerProxy` class already allows the outer to do that, minus the passing an iframe part (this whole example is currently assuming a single iframe and should be refactored to support multiple).  I definitely just left this here for now because I wasn't sure where it should go and iframe-driver didn't seem quite right.

---
In reply to: [540465462](https://github.com/microsoft/FluidFramework/pull/4404#discussion_r540465462) [](ancestors = 540465462)

anthony-murphy reviewed (approved):
:shipit:
